blacklist_test.go

```diff
@@ -0,0 +1,894 @@
+				{Type: BlacklistOpType_NodeSuspected, NodeIndex: 1},
+				{Type: BlacklistOpType_NodeRedeemed, NodeIndex: 2},
+				{Type: BlacklistOpType_NodeRedeemed, NodeIndex: 3},
+				{Type: BlacklistOpType_NodeRedeemed, NodeIndex: 4},
+				{Type: BlacklistOpType_NodeRedeemed, NodeIndex: 5},
+			},
+			expectedErr: errBlacklistTooManyUpdates,
+		},
+		{
+			name: "invalid type",
+			updates: []BlacklistUpdate{
+				{Type: 3, NodeIndex: 1},
+			},
+			expectedErr: errBlacklistInvalidUpdateType,
+		},
+		{
+			name: "invalid index",
+			updates: []BlacklistUpdate{
+				{Type: BlacklistOpType_NodeRedeemed, NodeIndex: 4},
+			},
+			expectedErr: errBlacklistInvalidNodeIndex,
+		},
+		{
+			name: "double vote",
+			updates: []BlacklistUpdate{
+				{Type: BlacklistOpType_NodeSuspected, NodeIndex: 3},
+				{Type: BlacklistOpType_NodeSuspected, NodeIndex: 2},
+				{Type: BlacklistOpType_NodeSuspected, NodeIndex: 3},
+			},
+			expectedErr: errBlacklistNodeIndexAlreadyUpdated,
+		},
+		{
+			name: "already blacklisted",
+			Blacklist: Blacklist{
+				NodeCount:      4,
+				SuspectedNodes: []SuspectedNode{{NodeIndex: 3, SuspectingCount: 2}},
+			},
+			updates: []BlacklistUpdate{
+				{Type: BlacklistOpType_NodeSuspected, NodeIndex: 3},
+				{Type: BlacklistOpType_NodeSuspected, NodeIndex: 2},
+			},
+			expectedErr: errBlacklistNodeAlreadyBlacklisted,
+		},
+		{
+			name: "not blacklisted, cannot be redeemed",
+			Blacklist: Blacklist{
+				NodeCount:      4,
+				SuspectedNodes: []SuspectedNode{{NodeIndex: 3, SuspectingCount: 1}},
+			},
+			updates: []BlacklistUpdate{
+				{Type: BlacklistOpType_NodeRedeemed, NodeIndex: 3},
+			},
+			expectedErr: errBlacklistNodeNotBlacklisted,
+		},
+	} {
+		t.Run(testCase.name, func(t *testing.T) {
+			err := testCase.Blacklist.verifyBlacklistUpdates(testCase.updates, 4)
+			require.ErrorContains(t, err, testCase.expectedErr.Error())
+		})
+	}
+}
+
+func TestBlacklistMultipleAttemptsOverSeveralOrbits(t *testing.T) {
+	// This test tests that even though a node is suspected continuously across several rounds,
+	// as long as the rounds are too far from each other, the node will not be blacklisted.
+
+	blacklist := Blacklist{
+		NodeCount: 4,
+	}
+
+	blacklist = blacklist.ApplyUpdates([]BlacklistUpdate{{Type: BlacklistOpType_NodeSuspected, NodeIndex: 3}}, 4)
+	require.False(t, blacklist.IsNodeSuspected(3))
+
+	blacklist = blacklist.ApplyUpdates([]BlacklistUpdate{{Type: BlacklistOpType_NodeSuspected, NodeIndex: 3}}, 7)
+	require.False(t, blacklist.IsNodeSuspected(3))
+
+	blacklist = blacklist.ApplyUpdates([]BlacklistUpdate{{Type: BlacklistOpType_NodeSuspected, NodeIndex: 3}}, 8)
+	require.True(t, blacklist.IsNodeSuspected(3))
+}
+
+func TestBlacklistSimulateNetwork(t *testing.T) {
+	for nodes := 1; nodes <= 128; nodes *= 2 {
+		for rounds := 10; rounds <= 1000; rounds *= 10 {
+			t.Run(fmt.Sprintf("n=%d/r=%d", nodes, rounds), func(t *testing.T) {
+				testBlacklistSimulateNetwork(t, nodes, rounds)
```

Comment:
we could run these in parallel for a speedup
```suggestion
                t.Parallel()
				testBlacklistSimulateNetwork(t, nodes, rounds)
```

Reply:
cuts the time in 50%